### PR TITLE
Update S104: "Too many lines in a file" should only count lines of code

### LINKED ISSUE #396: Update S104: "Too many lines in a file" should only count lines of code
[Rule S104](https://jira.sonarsource.com/browse/RSPEC-104)

The implementation of this rule should be adjusted to only take Lines of Code into account. This is part of [MMF-571](https://jira.sonarsource.com/browse/MMF-571), and also falls under the Kill the Noise initiative.

Also: https://stackoverflow.com/questions/44315302/s104-sonarqube-is-it-possible-that-the-summary-or-comments-are-not-added-in

### PR DESCRIPTION
**Pull request by Evangelink**
Fix #396

### CODE REVIEW

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Helpers/CSharpSyntaxHelper.cs [hunk 1]
```diff
@@ -1,252 +1,247 @@
-﻿/*
```
Comment (Evangelink) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Helpers/CSharpSyntaxHelper.cs]: Class rename.

File: sonaranalyzer-dotnet/src/SonarAnalyzer.VisualBasic/Helpers/VisualBasicSyntaxHelper.cs [hunk 2]
```diff
@@ -1,78 +1,78 @@
-﻿/*
```
Comment (Evangelink) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.VisualBasic/Helpers/VisualBasicSyntaxHelper.cs]: Class rename.

File: sonaranalyzer-dotnet/src/Tests/SonarAnalyzer.UnitTest/TestCases/FileLines20.cs [hunk 3]
```diff
@@ -0,0 +1,29 @@
+﻿// Noncompliant {{This file has 20 lines, which is greater than 10 authorized. Split it into smaller files.}}
```
Comment (duncanp-sonar) [inline sonaranalyzer-dotnet/src/Tests/SonarAnalyzer.UnitTest/TestCases/FileLines20.cs]: You could amend this test file to make the tests slightly more comprehensive with:
* a block comment / doc comment, and 
* a line that contains both code and a comment

Out of interest, do you know if a completely empty file has a syntax tree?

Comment (Evangelink) [reply sonaranalyzer-dotnet/src/Tests/SonarAnalyzer.UnitTest/TestCases/FileLines20.cs]: I have added those cases onto the `FileLines9.cs` test case.

File: sonaranalyzer-dotnet/src/SonarAnalyzer.Common/Helpers/SyntaxHelper.cs [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+using Microsoft.CodeAnalysis;
+
+namespace SonarAnalyzer.Helpers
+{
+    internal static class SyntaxHelper
+    {
+        public static int GetLineNumber(this SyntaxToken token, bool isZeroBasedCount = true)
```
[the remark this answers is not in the document]
Comment (Evangelink) [reply sonaranalyzer-dotnet/src/SonarAnalyzer.Common/Helpers/SyntaxHelper.cs]: It is not only SonarQube but all tools that uses the 1-based line approach for files. I was tempted to have the 2 methods at first but I though that wasn't better.

[the remark this answers is not in the document]
Comment (duncanp-sonar) [reply sonaranalyzer-dotnet/src/SonarAnalyzer.Common/Helpers/SyntaxHelper.cs]: The method is internal to _SonarAnalyzer.Common_. Which other tools would we use it with?

[the remark this answers is not in the document]
Comment (Evangelink) [reply sonaranalyzer-dotnet/src/SonarAnalyzer.Common/Helpers/SyntaxHelper.cs]: Sorry what I meant by tool is everything is displaying files with a 1-based line approach (VS text editor, GitHub....) so it seems more natural to use it than a 0-based one. As the existing code base was calling the `GetLineNumber()` and expecting a zero-based I set the extra param as true by default.

File: sonaranalyzer-dotnet/src/Tests/SonarAnalyzer.UnitTest/TestCases/FileLines20.cs (hunk 3 above)
Comment (duncanp-sonar) [reply sonaranalyzer-dotnet/src/Tests/SonarAnalyzer.UnitTest/TestCases/FileLines20.cs]: It's not a big deal, but it isn't possible to tell from the sample code in _FileLines9.cs_ what happens when a line has both code and a comment. The line might or might not be counted, but either way the LOC is still well under the maximum. The only way to check that lines that contain code and a comment are correctly counted would be with a non-compliant example, where the (code + comment) line is included in the LOC reported for the file.

**Pull request merged**